### ISSUE TITLE
Feature: batched + warm-started differentiable solve in pounce.jax for inner-loop optimization

### ISSUE THREAD

**Issue by robertojjx** [jkitchin/pounce#74]
## Summary

Two related enhancements that would make `pounce.jax` usable as the **inner solver for batched, iterative differentiable optimization** (e.g. a constrained-projection layer inside a neural-network training loop). Both behaviors are currently **by design and documented** — this is a feature request, not a bug report.

### 1. A batched differentiable solve (not a sequential loop)

`pounce.jax.vmap_solve` runs `jax.lax.map` — one solve at a time — because the solver is single-threaded and stateful (`__init__.py` and `notebooks/03_implicit_differentiation.ipynb` both note this, and a literal `jax.vmap` would unsafely lift the impure `pure_callback`).

For use cases that project/solve a **whole minibatch every step**, this is the dominant cost: `N` independent solves dispatched sequentially, each rebuilding its `IpoptApplication`. A batched entry point — a pool of reusable solver instances, or the Rust-side batched sensitivity solve hinted at in `_diff.py` ("the right move is to expose a Rust-side sensitivity-solve via `pounce-sensitivity`; that lands as a follow-up") — would remove the structural disadvantage.

### 2. Warm-start across differentiable solves

`pounce.jax.solve` is documented as not differentiable w.r.t. `x0`, and threads no warm primal/dual state between calls. In an iterative outer loop the parameter `p` moves only slightly per step, so the previous primal+dual solution is an excellent warm start — but there's currently no way to carry it across `solve` calls through the JAX boundary, so every step pays a cold interior-point solve.

A request: an opt-in to pass and recover warm primal/dual state across successive `solve` calls (without making `x0` differentiable — the warm start need not participate in the gradient).

## Why these two together

A differentiable constrained layer does, per training step: solve `N` projections, each warm-started from the last epoch's solution, and backprop through all of them. Today that maps onto `pounce.jax` as `N` cold, sequential solves — correct, but slow relative to a backend that batches and warm-starts. Either enhancement helps; both together would make `pounce.jax` competitive as an inner solver for this class of problem.

## Acknowledgement

Both are reasonable defaults given the stateful Rust core and the black-box `pure_callback` design — filing as enhancement requests, with the use case spelled out so the priority can be judged against it. Environment: POUNCE 0.2.0 at `6be3643`, jax 0.6.2.

**Comment by jkitchin** [comment]
Thanks for the careful framing — both the use case ("inner solver for batched, iterative differentiable optimization") and the pointers into `_diff.py` / `notebooks/03_implicit_differentiation.ipynb` made this easy to size against.

Splitting the response by request because the two land very differently:

## 2. Warm-start across differentiable solves — **shipped in 0.3.0**

This was the cheap one and the one with the higher leverage per line of code for your stated use case (small `p` step → previous duals are nearly correct → cold IPM is pure waste). It's now on `main` in 6f30b27 as a new top-level export, `pounce.jax.solve_with_warm`:

```python
from pounce.jax import solve_with_warm

x_star, warm = solve_with_warm(
    p, f=f, g=g, x0=x0, n=n, m=m,
    lb=lb, ub=ub, cl=cl, cu=cu,
    warm_start=None,                # cold first call
)
for p_k in trajectory[1:]:
    x_star, warm = solve_with_warm(
        p_k, f=f, g=g, x0=x_star, n=n, m=m,
        lb=lb, ub=ub, cl=cl, cu=cu,
        warm_start=warm,            # threaded dual triple (mult_g, mult_x_L, mult_x_U)
    )
```

The threaded `warm` triple is `(mult_g, mult_x_L, mult_x_U)` — exactly the dual state IPOPT consumes when `warm_start_init_point=yes` is set. `solve_with_warm` sets that option for you (you can still override it in `options=` if you need different `warm_start_bound_push` / `warm_start_mult_bound_push` defaults).

**On differentiability:** `solve_with_warm` is differentiable w.r.t. `p` only — cotangents on the warm-state outputs and the warm-state inputs are dropped (zero), matching how `solve` already handles `x0`. The implicit-function-theorem argument is the same: at the optimum the duals are a function of `p` and the active set, not an independent input that feeds into `dx*/dp`. So the warm-state plumbing is *pure compute-saving* — it has zero effect on the gradient you get out of `jax.grad(loss)(p)`. (This was the thing I most wanted to verify before pushing — the regression test in `python/tests/test_jax.py::test_solve_with_warm_reduces_iterations_pounce_74` pins it on the same box-projection fixture as the #73 tests.)

**Non-breaking:** `solve` and `vmap_solve` are untouched. If you're already on the `solve()` surface and just want to swap one call site at a time, the migration is mechanical:

```diff
- x_star = solve(p, f=f, g=g, x0=x0, n=n, m=m, ...)
+ x_star, warm = solve_with_warm(p, f=f, g=g, x0=x0, n=n, m=m, ..., warm_start=warm)
```

(Initial `warm=None` cold-starts; `x0` keeps doing primal warm-start as before.)

## 1. A truly batched differentiable solve — **bigger lift, not in 0.3.0**

Sizing honestly: this is a multi-week piece of work, not a same-day patch, and `pounce` 0.3.0 is about to tag with the warm-start fix above. The blocker is structural and I want to lay it out so the priority can be judged:

**Why naive `jax.vmap` is unsafe today.** `pounce.jax.solve` goes through `jax.pure_callback` into a Python host function that owns a Rust `IpoptApplication`. The Rust side is stateful per problem (option stash, KKT factorization cache, working-set warm-start payload), so two `vmap`-lifted callbacks racing into the same `Problem` would corrupt internal state.

**Why a Python `concurrent.futures` threadpool wouldn't help either.** I confirmed by grep over `crates/pounce-py/src/problem.rs::Problem::solve` that there is no `py.allow_threads(...)` block on the `optimize_tnlp` path — the GIL is held for the full duration of the solve. So even if we lifted the callback into a threadpool, the GIL would serialize the solves into the exact `lax.map` behavior `vmap_solve` already has. The threadpool option is a non-starter until `pounce-py` releases the GIL across the solve call (one-line PyO3 change in principle, but it has to be audited against every callback the bridge makes back into Python for `f` / `g` / Hessian — those callbacks must re-acquire the GIL, and JAX-traced callables in particular need careful handling).

**The two real paths forward, in order of effort:**

1. **`py.allow_threads` around `optimize_tnlp` + a Python-side pool of independent `Problem` instances.** Each worker holds its own `Problem`, so the per-instance state is no longer shared. Calls become genuinely concurrent on the GIL, and `vmap_solve` becomes a real parallel batch (limited by core count and BLAS thread affinity). Forward only — no JAX-level batching of the backward, but the backward is JAX-native and already vectorizes cleanly via `vmap` over the per-element KKT solve. Rough sizing: ~1 week including the callback-reentrancy audit on the Python AD path and a regression test that confirms two threads running `pounce.jax.solve` on independent `Problem`s produce identical numerics to the sequential reference. The gain depends entirely on how compute-bound each individual solve is — for the small constrained-projection problems in your use case (`n` and `m` in the tens-to-hundreds), I'd expect ~2–4× on a typical 8-core laptop before BLAS contention kicks in.

2. **A Rust-side batched sensitivity solve via `pounce-sensitivity`** (the "follow-up" hinted at in `_diff.py`). This batches the *backward* — every element of the batch hits the same KKT factorization structure, so a single factorize + N back-solves is genuinely a win when `N` is large. Useful but orthogonal to (1) — the *forward* is still N independent solves; this only speeds the implicit-diff. Heavier lift: ~2–3 weeks because `pounce-sensitivity` doesn't yet expose a Python-callable batched solve surface, and the JAX side needs a new custom_vjp entry point that calls into it.

For the inner-solver-in-a-training-loop use case you describe, I'd guess (1) gives you the bigger practical speedup — the forward dominates, and warm-start (now shipped) already cuts the per-element forward cost. Combining `solve_with_warm` with parallel `vmap_solve` would land most of the win you're looking for.

**My ask:** would you be willing to share a representative benchmark (your real constrained-projection layer at training-relevant batch size, even just a single forward+backward pass at fixed `p`), so the speedup target for (1) can be measured against actual workload rather than synthetic timing? That'd also confirm whether the GIL-release path is enough, or whether you'd in practice be bottlenecked by something else (callback overhead, JIT recompilation across batch elements, etc.) that would change the design.

In the meantime: please give `solve_with_warm` a spin against your trajectory — the API is documented in the docstring (`python -c "from pounce.jax import solve_with_warm; help(solve_with_warm)"`) and there's a worked example in the docstring's "Typical use" block. Anything that doesn't match your expectations, please reopen / comment here.

**Comment by jkitchin** [comment]
Quick follow-up — #74-1 also shipped in c248598 (on `main`), faster than the ~1 week I sized this at last week because the bridge audit went better than expected. Both halves of #74 are now in 0.3.0.

## What landed

Two changes that compose:

### 1. `pounce_py::Problem::solve` now releases the GIL across `optimize_tnlp`

The audit I mentioned in the previous comment went well: every TNLP callback in `crates/pounce-py/src/tnlp_bridge.rs` (`eval_f`, `eval_grad_f`, `eval_g`, `eval_jac_g`, `eval_h`, `intermediate_callback`) was already wrapped in `Python::with_gil(...)` before touching `Py<PyAny>` — the bridge was designed with eventual GIL release in mind, the actual `py.allow_threads` was just never wired up. So this is a localized change in `crates/pounce-py/src/problem.rs`:

```rust
let (status, app_back, bridge_back) = py.allow_threads(move || {
    let mut app = app_guard.into_inner();
    let bridge = bridge_guard.into_inner();
    let bridge_for_solve: Rc<RefCell<dyn TNLP>> = bridge.clone();
    let status = app.optimize_tnlp(bridge_for_solve);
    (status, SendGuard::new(app), SendGuard::new(bridge))
});
```

The `SendGuard` is an `unsafe impl Send` shim. The safety argument is that `Python::allow_threads` runs its closure on the *same* OS thread — `PyEval_SaveThread` releases the GIL but doesn't migrate the thread — so the `Rc` refcount is only ever touched by one thread despite the move. That's reasoned about in a docstring on the block.

### 2. `pounce.jax.vmap_solve_parallel` — drop-in for `vmap_solve`

```python
from pounce.jax import vmap_solve_parallel

x_batch = vmap_solve_parallel(
    p_batch, f=f, g=g, x0=x0, n=n, m=m,
    lb=lb, ub=ub, cl=cl, cu=cu,
    workers=8,  # default: min(B, 8)
)
```

Each of the `B` batch elements is dispatched to a worker in a `ThreadPoolExecutor`. Each worker builds and owns its own `Problem` (no shared state). The forward is parallel via the threadpool; the backward is `jax.vmap` over the per-element KKT solve, which is pure JAX and vectorizes naturally.

## Speedup

Microbench on the M1 8-core machine I happen to have, `n = 30`, `B = 16`, nonlinear unconstrained `f(x, p) = ||x - p||² + 0.01 ||x||⁴`:

```
workers=1: 0.942s  speedup=1.07x  max|x_serial - x_par|=0.00e+00
workers=2: 0.480s  speedup=2.09x  max|x_serial - x_par|=0.00e+00
workers=4: 0.406s  speedup=2.47x  max|x_serial - x_par|=0.00e+00
workers=8: 0.365s  speedup=2.75x  max|x_serial - x_par|=0.00e+00

serial (vmap_solve): 1.004s  on B=16, n=30
```

(`max|x_serial - x_par| = 0` is exact — every batch element converges to the same optimum to the limits of the solver's `tol=1e-9`.)

2.75× at 8 workers on a small problem. The scaling falls off because:
1. `n=30` is small enough that the Python callback re-entry for `f` and `g` is a non-trivial fraction of per-element time — those *do* serialize on the GIL even with the release in place.
2. BLAS thread contention starts mattering past 4 workers on M1.

I'd expect closer to linear scaling on bigger problems where the per-element solve is heavily compute-bound. **This is where your benchmark would really help** — if you can share even a single forward+backward at training-relevant batch size and `n`, I can confirm the speedup ratio on the actual workload and tune the default `workers` accordingly.

## Combined with #74-2

`vmap_solve_parallel` and `solve_with_warm` compose — for the typical inner-loop use case, you'd want both. The "right shape" is roughly:

```python
# initial cold parallel solve over the batch
x_batch = vmap_solve_parallel(p_batch_0, f=f, g=g, x0=x0_batch, ...)
warm_batch = (lam_batch, zL_batch, zU_batch)  # collected from the first pass

for p_batch_k in trajectory[1:]:
    # warm-started parallel solve — each worker uses its own slice of the dual triple
    x_batch, warm_batch = vmap_solve_parallel_warm(p_batch_k, x0=x_batch, warm=warm_batch, ...)
```

Note that *combined* surface (`vmap_solve_parallel_warm`) doesn't exist yet — I held off because the API question of "is the warm-state per-batch-element or shared?" deserves a real use case to decide against. The two ingredients are both there now; happy to compose them once your workload tells me which way the warm-state plumbing should go.

## Test coverage

Two new regression tests in `python/tests/test_jax.py`:

- `test_vmap_solve_parallel_matches_vmap_solve_pounce_74` — unconstrained, B=4, n=3. Asserts forward equals serial reference *and* `jax.grad` through the parallel surface matches the closed-form `2 p`.
- `test_vmap_solve_parallel_with_constraints_pounce_74` — mixed active/inactive inequality across batch elements. Confirms per-element active-set logic is respected when the GIL is released and JAX callbacks re-enter.

All 7 JAX tests pass at c248598.

## Closing the issue

Closing #74 — both halves shipped. If anything misbehaves in the actual workload, please reopen with the failing case (especially: a benchmark that exercises the parallel speedup on your problem size, or anything where `vmap_solve_parallel` produces different gradients than `vmap_solve`).

**Comment by jkitchin** [comment]
Closing per the resolution comment above — both halves of the request (warm-start and parallel batched solve) shipped in 0.3.0.
